### ISSUE TITLE
FO4Edit: Array Sorting

### ISSUE THREAD

**Issue by fireundubh** [TES5Edit/TES5Edit#294]
OT: Can we rename Component's wbFormIDCk field to Form? Makes more sense and less confusing.

# MISC

In `MISC`, you wrote:

> the amount of components is the same as size of CDIX, so should not be sorted probably

This works fine though. No errors.

```
    wbArrayS(CVPA, 'Components',
      wbStructSK([0], 'Component', [
        wbFormIDCk('Form', [CMPO, MISC, ALCH, AMMO]),
        wbInteger('Count', itU32)
      ])
    ),
```

`CDIX` also looks like flags, not a components count.

CDIX | Form ID
--- | ---
`00 00 03` | `DeskFanOfficeOff01 [MISC:00192D5D]`
`00 01` | `OilCan01 [MISC:0004B273]`
`00 01` | `OilCan01_PreWar [MISC:000FCE33]`
`00 01` | `PrewarTincan01 [MISC:000F15BB]`
`00 02 01` | `FireExtinguisher01 [MISC:0001F8F9]`
`00` | `DN102_HalluciGenGasCanister [MISC:000E69AF]`
`01` | `CoffeeCup01 [MISC:000211C4]`
`01` | `CoffeeCup01clean [MISC:00171947]`
`01` | `COMMacCreadyToySolider [MISC:00191E93]`
`01` | `LukowskisPottedMeatEmpty [MISC:0016D8C9]`
`01` | `LukowskisPottedMeatUnfilled [MISC:0016D8CA]`
`01` | `MoleRatTeeth [MISC:00034603]`
`01` | `TinCan01 [MISC:0001F90A]`
`01` | `TinCan02 [MISC:0001F908]`
`03 01 03` | `DeskFanOffice01 [MISC:0001F905]`
`03 01 03` | `PrewarDeskFanOffice01 [MISC:000F15BF]`

# COBJ

Sorting categories is helpful for conflict discovery:

`wbArrayS(FNAM, 'Category', wbFormIDCk('Keyword', [KYWD])),`

**Comment by zilav** [comment]
I have a suspicion that the order of elements in CDIX must correlate with CVPA, that's why it is unsorted.
I'm also not sure if the order of COBJ categories affects positions of those categories in ingame crafting menues, so left them unsorted for now. You can check that if you wish or I'll just wait for CK.

**Comment by fireundubh** [comment]
> I have a suspicion that the order of elements in CDIX must correlate with CVPA, that's why it is unsorted.

I don't know what that means. The table above lists all of the MISC records that have CDIX assigned, and none of the CDIX values in those few records correspond to CVPA except by happenstance.

> I'm also not sure if the order of COBJ categories affects positions of those categories in ingame crafting menues, so left them unsorted for now. You can check that if you wish or I'll just wait for CK.

There's only one COBJ record that has more than one FNAM keyword:

### workshop_co_FastTravelTarget [COBJ:001BDDB8]

* `WorkshopRecipeFilterResource02Misc "Miscellaneous" [KYWD:00118F64]`
* `WorkshopAlwaysShowIconKeyword [KYWD:00237B63]`

Game doesn't use FNAM to order recipes in the crafting menus. The menus themselves are ordered by the respective `WorkshopMenu*` menu formlists while the objects are sorted by `INTV\Unknown`.

You could name `INTV\Unknown` "Menu Order" or "Priority."

INTV\Unknown | Value | Form ID
--- | --- | ---
`1E 00` | 30 | `workshop_co_TurretTripod [COBJ:00076AB9]`
`20 00` | 32 | `workshop_co_TurretTripodHeavy [COBJ:0022B59F]`
`28 00` | 40 | `workshop_co_TurretLaser [COBJ:000D34D2]`
`28 00` | 40 | `workshop_co_TurretShotgun [COBJ:0022B175]`
`28 00` | 40 | `workshop_co_TurretSpotlight [COBJ:0022B18E]`
`2A 00` | 42 | `workshop_co_TurretLaserHeavy [COBJ:0022B59E]`
`64 00` | 100 | `workshop_co_TurretMissile [COBJ:0022B176]`

These values correspond to the order of these turrets in the Defense/Turrets menu.

1. Machinegun Turret
2. Heavy Machinegun Turret
3. Laser Turret
4. Shotgun Turret
5. Spotlight
6. Heavy Laser Turret
7. Missile Turret

The three turrets with the same priority are sorted alphabetically.

Also "Filter" is probably more accurate than "Category."

**Comment by zilav** [comment]
The size of CDIX is always equal to the number of components, that's why I think they correlate.

**Comment by fireundubh** [comment]
> The size of CDIX is always equal to the number of components, that's why I think they correlate.

Seems you're onto something! But what!

CVPA Count | CDIX Bytes | Form ID
--- | --- | ---
1 | 1 | CoffeeCup01 [MISC:000211C4]
1 | 1 | CoffeeCup01clean [MISC:00171947]
1 | 1 | COMMacCreadyToySolider [MISC:00191E93]
1 | 1 | DN102_HalluciGenGasCanister [MISC:000E69AF]
1 | 1 | LukowskisPottedMeatEmpty [MISC:0016D8C9]
1 | 1 | LukowskisPottedMeatUnfilled [MISC:0016D8CA]
1 | 1 | MoleRatTeeth [MISC:00034603]
1 | 1 | TinCan01 [MISC:0001F90A]
1 | 1 | TinCan02 [MISC:0001F908]
2 | 2 | OilCan01 [MISC:0004B273]
2 | 2 | OilCan01_PreWar [MISC:000FCE33]
2 | 2 | PrewarTincan01 [MISC:000F15BB]
3 | 3 | DeskFanOffice01 [MISC:0001F905]
3 | 3 | DeskFanOfficeOff01 [MISC:00192D5D]
3 | 3 | FireExtinguisher01 [MISC:0001F8F9]
3 | 3 | PrewarDeskFanOffice01 [MISC:000F15BF]

**Comment by fireundubh** [comment]
I think `CDIX` is an array of priorities, where a higher number is a higher priority, for determining which items to scrap first when you don't have enough raw components.

I removed all items from my person and workbench with `RemoveAllItems`, and gave myself:

* 10x `OilCan01 [MISC:0004B273]`
* 10x `PrewarDeskFanOffice01 [MISC:000F15BF]`

Then I built `workshop_co_HighTechAshtray_01_Dirty [COBJ:002381AC]`, which required 4x Steel.

The game decided to scrap 2x `PrewarDeskFanOffice01 [MISC:000F15BF]` to make 4x Steel available.

Each subsequent Ashtray resulted in two less `PrewarDeskFanOffice01 [MISC:000F15BF]`.

Needs more testing to verify.

### OilCan01 [MISC:0004B273]

CDIX | Qty | Component
--- | --- | ---
`00` | `1` | `c_Steel "Steel" [CMPO:0001FABD]`
`01` | `1` | `c_Oil "Oil" [CMPO:0001FAB4]`

### PrewarDeskFanOffice01 [MISC:000F15BF]

CDIX | Qty | Component
--- | --- | ---
`03` | `2` | `c_Steel "Steel" [CMPO:0001FABD]`
`01` | `2` | `c_Screws "Screw" [CMPO:0003D294]`
`03` | `2` | `c_Gears "Gear" [CMPO:0001FAB0]`

**Comment by zilav** [comment]
Good to know, but this means that Components can't be sorted, at least when CDIX is present.

**Comment by Hugues92** [comment]
if it exists, CDIX fills the beginning of an array of the same allocation (memory allocated div by element size) as the one containing the values from CVPA. That array does not exists if CDIX does not exists.

So "logic dictates" it is an array of bytes. The count should not be greater than the one from CVPA.

How about it is the perk level required to recover said component when scrapping? (I don't remember the perk name right now).

**Comment by Hugues92** [comment]
By the way, DAMC in Destructible uses the same routine to load itself as CVPA, so they should support the same definition.

**Comment by fireundubh** [comment]
> How about it is the perk level required to recover said component when scrapping? 

Scrapper has only two ranks whereas there are four used values in CDIX (0, 1, 2, 3).

**Comment by fireundubh** [comment]
@zilav: I did another test and changed the CDIX values for the previously tested forms. There was no effect. Unfortunately, I think we can scratch off the autoscrap priority hypothesis.

I also changed the full name of `PrewarDeskFanOffice01 [MISC:000F15BF]` to determine whether autoscrap was using alphabetical order to prioritize items, but that, too, had no effect.

So, I have no idea what CDIX means, and now I have no idea how autoscrap prioritizes items.
